File: algorithms/2_medium/J/01696_jump_game_vi/heap.cpp
class Solution {
public:
    static int maxResult(const vector<int>& nums, const int k) {
        assert(nums.size() >= 1 && nums.size() <= 10e5 && k >= 1);

        if (nums.size() == 1 || k == 1) {
            return reduce(nums.cbegin(), nums.cend(), 0);
        }

        const int32_t len = static_cast<int32_t>(std::ssize(nums));
        auto heap = priority_queue<pair<int, int32_t>>();
        heap.push({nums[0], 0});
        int dp_i = nums[0];

        for (int32_t i = 1; i < len; i += 1) {
            const int32_t start = std::max(i - k, 0);

            while (heap.top().second < start) {
                heap.pop();
            }

            dp_i = nums[i] + heap.top().first;
            heap.push({dp_i, i});
        }

        return dp_i;
    }
};

File: algorithms/2_medium/J/01696_jump_game_vi/heap.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn max_result(nums: Vec<i32>, k: i32) -> i32 {
        assert!(nums.len() >= 1 && nums.len() <= 100_000 && k >= 1);

        if nums.len() == 1 || k == 1 {
            return nums.iter().sum();
        }

        let len = nums.len() as u32;
        let k = k as u32;
        let mut heap = BinaryHeap::with_capacity(len as usize);
        heap.push((nums[0], 0));
        let mut dp_i = nums[0];

        for i in 1..len {
            let start = if i >= k { i - k } else { 0 };

            while let Some(top) = heap.peek() && top.1 < start {
                heap.pop();
            }

            dp_i = nums[i as usize] + heap.peek().unwrap().0;
            heap.push((dp_i, i));
        }

        dp_i
    }
}
